stlab/concurrency/main_executor.hpp

```diff
@@ -104,7 +104,28 @@ struct main_executor_type {
 
 #elif STLAB_MAIN_EXECUTOR(EMSCRIPTEN)
 
-using main_executor_type = default_executor_type;
+struct main_scheduler_type {
+    using result_type = void;
+
+    template <class F>
+    void operator()(F&& f) const {
+        using function_type = typename std::remove_reference<F>::type;
+        auto p = new function_type(std::forward<F>(f));
+
+        emscripten_async_run_in_main_runtime_thread(
```

Comment:
Can you point me to where the documentation for this is?

Reply:
Keeping the comments from the original code would probably be useful (scrubbed):
```
        /*
           `emscripten_async_run_in_main_runtime_thread()` schedules a function to run on the main
            JS thread, however, the code can be executed at any POSIX thread cancelation point if
            wasm code is executing on the JS main thread.
            Executing the code from a POSIX thread cancelation point can cause problems, including
            deadlocks and data corruption. Consider:
            ```
                mutex.lock();   // <-- If reentered, would deadlock here
                new T;          // <-- POSIX cancelation point, could reenter
            ```
            The call to `emscripten_async_call()` bounces the call to execute as part of the main
            run-loop on the current (main) thread. This avoids nasty reentrancy issues if executed
            from a POSIX thread cancelation point.
        */
```

Documentation for the call is in the header https://chromium.googlesource.com/external/github.com/kripken/emscripten/+/refs/tags/1.37.5/system/include/emscripten/threading.h - however, it is incomplete. This incantation was worked out with a chat with Google.

Reply:
Apologies for leaving the comment behind. Just added it back.